server/repository/src/db_diesel/assets/asset_row.rs

```diff
@@ -0,0 +1,100 @@
+use crate::StorageConnection;
+
+use chrono::NaiveDate;
+use chrono::NaiveDateTime;
+use diesel::prelude::*;
+
+table! {
+    asset (id) {
+        id -> Text,
+        store_id -> Text,
+        serial_number -> Text,
+        asset_category_id -> Text,
+        asset_type_id -> Text,
+        catalogue_item_id -> Nullable<Text>,
+        installation_date -> Nullable<Date>,
+        replacement_date -> Nullable<Date>,
+        created_datetime -> Timestamp,
+        modified_datetime -> Timestamp,
+        deleted_datetime -> Nullable<Timestamp>,
+    }
+}
+
+#[derive(Clone, Insertable, Queryable, Debug, PartialEq, AsChangeset, Eq, Default)]
+#[table_name = "asset"]
+pub struct AssetRow {
+    pub id: String,
+    pub store_id: String,
+    pub serial_number: String,
+    #[column_name = "asset_category_id"]
+    pub category_id: String,
+    #[column_name = "asset_type_id"]
+    pub type_id: String,
+    pub catalogue_item_id: Option<String>,
+    pub installation_date: Option<NaiveDate>,
+    pub replacement_date: Option<NaiveDate>,
+    pub created_datetime: NaiveDateTime,
+    pub modified_datetime: NaiveDateTime,
+    pub deleted_datetime: Option<NaiveDateTime>,
+}
+
+pub struct AssetRowRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> AssetRowRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        AssetRowRepository { connection }
+    }
+
+    #[cfg(feature = "postgres")]
+    pub fn upsert_one(&self, asset_row: &AssetRow) -> Result<(), RepositoryError> {
+        diesel::insert_into(asset)
+            .values(asset_row)
+            .on_conflict(id)
+            .do_update()
+            .set(asset_row)
+            .execute(&self.connection.connection)?;
+        Ok(())
+    }
+
+    #[cfg(not(feature = "postgres"))]
+    pub fn upsert_one(&self, asset_row: &AssetRow) -> Result<(), RepositoryError> {
+        diesel::replace_into(asset)
+            .values(asset_row)
+            .execute(&self.connection.connection)?;
+        Ok(())
+    }
+
+    pub fn insert_one(&self, asset_row: &AssetRow) -> Result<(), RepositoryError> {
+        diesel::insert_into(asset)
+            .values(asset_row)
+            .execute(&self.connection.connection)?;
+        Ok(())
+    }
+
+    pub fn find_all(&self) -> Result<Vec<AssetRow>, RepositoryError> {
+        let result = asset
+            .filter(deleted_datetime.is_null())
+            .load(&self.connection.connection);
+        Ok(result?)
+    }
+
+    pub fn find_one_by_id(&self, asset_id: &str) -> Result<Option<AssetRow>, RepositoryError> {
+        let result = asset
+            .filter(id.eq(asset_id))
+            .first(&self.connection.connection)
+            .optional()?;
+        Ok(result)
+    }
+
+    pub fn delete(&self, asset_id: &str) -> Result<(), RepositoryError> {
+        diesel::update(asset.filter(id.eq(asset_id)))
+            .set(deleted_datetime.eq(Some(chrono::Utc::now().naive_utc())))
```

Comment:
Implemented soft delete for assets, seems easiest to do it now rather than later...